File: py/altschul.py
"""Explore permutations.

References:

Significance of nucleotide sequence alignments...
- https://academic.oup.com/mbe/article/2/6/526/981780
Shuffling biological sequences
- https://www.sciencedirect.com/science/article/pii/S0166218X97814564
Shuffling biological sequences with motif constraints
- https://www.sciencedirect.com/science/article/pii/S1570866707000548
"""

import random
from collections import defaultdict
from pprint import pprint

from constants import S1
from utils import UnionFind, chunkify_cons, same_klets, same_klons

SEP = "==" * 20


def check_if_connected(z_graph: list[str], vertices: list[str]) -> bool:
    """Check if all last edges are connected to sf."""
    assert len(z_graph) == len(vertices) - 1

    uf = UnionFind(len(vertices))
    for edge in z_graph:
        fr = edge[:-1]
        to = edge[1:]
        fr_idx = vertices.index(fr)
        to_idx = vertices.index(to)
        uf.union(fr_idx, to_idx)

    return uf.group_count() == 1


def edge_ordering(seq: str, k: int) -> dict[str, list[str]]:
    eord = defaultdict(list)
    for chunk in chunkify_cons(seq, k):
        eord[chunk[: k - 1]].append(chunk)

    # The last vertex must be added if it previously was not.
    eord[seq[-(k - 1) :]].extend([])

    # The sorting is not needed: for visualization
    eord = dict(sorted(eord.items()))
    return eord


def doublet_preserving_permutation(seq: str, k: int = 2, *, debug: bool = False) -> str:
    assert k < len(seq)

    # 1. Construct the klet graph G and edge ordering E(S)
    eord = edge_ordering(seq, k)
    s1 = seq[: k - 1]
    sf = seq[-(k - 1) :]

    if debug:
        print("1.", SEP)
        print(seq)
        pprint(eord)
        print(f"{s1=}, {sf=}")

    # 2. For each vertex in G \ s_f, randomly select last edges
    # 3. From the set of last edges, construct z_graph
    #    and assess if connected to s_f
    # 4. If it is not connected, goto (2.)
    last_edges = []
    vertices = list(eord.keys())
    seen = set()

    MAX_ITERATIONS = 100
    while MAX_ITERATIONS > 0:
        last_edges.clear()
        for vertex, edges in eord.items():
            if vertex != sf:
                last_edges.append(random.choice(edges))

        key = tuple(last_edges)
        if key in seen:
            continue
        seen.add(key)

        if check_if_connected(last_edges, vertices):
            break

        MAX_ITERATIONS -= 1

    if not MAX_ITERATIONS:
        original_last_edges = [edges[-1] for edges in eord.values() if edges]
        # We got unlucky and did not even find the original endings once!
        if tuple(original_last_edges) not in seen:
            # msg = f"Exhausted iterations.\nReturning original sequence."
            # print(msg)
            assert False
            return seq
        else:
            msg = f"Exhausted iterations even though we found the original sequence.\n{eord=}\n{seen=}\n{original_last_edges=}\n\nReturning original sequence."
            raise RuntimeError(msg)

    if debug:
        print("5.", SEP)
        print(f"{last_edges=}")

    # Put last edges at the end
    for last_edge in last_edges:
        vertex = last_edge[: k - 1]
        last_edge_idx = eord[vertex].index(last_edge)
        eord[vertex].pop(last_edge_idx)
        eord[vertex].append(last_edge)

    # 5. Randomly permute:
    #    - every edge for s_f
    #    - every edge but the last edge chose for G \ s_f
    def permute_except_last(lst: list[str]) -> list[str]:
        to_permute = lst[:-1]
        random.shuffle(to_permute)
        return to_permute + [lst[-1]]

    eord_perm = dict()
    for vertex, edges in eord.items():
        if vertex == sf:
            eord_perm[vertex] = edges
            random.shuffle(eord_perm[vertex])
        else:
            perm_edges = permute_except_last(edges)
            eord_perm[vertex] = perm_edges

    # 6. Construct the sequence from the permuted edge ordering
    new_seq = s1
    cur_vertex = s1
    if debug:
        print(k)
        print(cur_vertex, None, new_seq)
    while True:
        try:
            cur_edge = eord_perm[cur_vertex].pop(0)
        except IndexError:
            assert cur_vertex == sf
            # print(f"Exiting gracefully: {cur_vertex} was not in edge ordering")
            break
        cur_vertex = cur_edge[1:]
        new_seq += cur_edge[-1]
        if debug:
            print(cur_edge, cur_vertex, new_seq)

    for key, value in eord_perm.items():
        if value:
            msg = f"{key} edge list was not exhausted\n{seq=}\n{last_edges=}"
            raise RuntimeError(msg)
    if len(seq) != len(new_seq):
        msg = f"Expected new_seq to be of len {len(seq)} but was {len(new_seq)}"
        raise RuntimeError(msg)
    if not same_klets(seq, new_seq, 2):
        msg = "Doublets are not preserved"
        raise RuntimeError(msg)

    return new_seq


# DTP
def doublet_and_triplet_preserving_permutation(seq: str) -> str:
    new_seq = doublet_preserving_permutation(seq, k=3)

    if not same_klets(seq, new_seq, 2):
        msg = "Doublets are not preserved"
        raise RuntimeError(msg)
    if not same_klets(seq, new_seq, 3):
        msg = "Triplets are not preserved"
        raise RuntimeError(msg)

    return new_seq


# DtP
def doublet_and_triplon_preserving_permutation(seq: str, *, debug: bool = True) -> str:
    if len(seq) % 3:
        print("Could not DtP permutate, the sequence len is not a multiple of 3")
        return seq

    # 1.
    seq = seq.upper()

    # 2.
    s_star = "".join(seq[i : i + 2] + seq[i + 2].lower() for i in range(0, len(seq), 3))
    if debug:
        print(f"s ={seq}")
        print(f"s*={s_star}")

    # 3.
    eord = defaultdict(list)
    for chunk in chunkify_cons(s_star, 3):
        eord[chunk[0] + chunk[2]].append(chunk)
    # The sorting is not needed: for visualization
    eord = dict(sorted(eord.items()))
    if debug:
        pprint(eord)

    # 4. Create r
    r = "".join(s_star[i] + s_star[i + 2] for i in range(0, len(s_star), 3))

    # 5. permute r
    r_perm = doublet_preserving_permutation(r)
    if debug:
        print(f"r ={r}")
        print(f"r'={r_perm}")

    # 6.
    for _, edges in eord.items():
        random.shuffle(edges)

    new_seq = ""
    for idx in range(0, len(r_perm), 2):
        cur_doublet = r_perm[idx : idx + 2]
        try:
            cur_triplon = eord[cur_doublet].pop(0)
        except IndexError:
            print(f"Could not find {cur_doublet} in eord")
            break

        new_seq += cur_triplon
        # if debug:
        #     print("> ", r_perm, cur_doublet, "=>", cur_triplon)

    new_seq = new_seq.upper()

    # NOTE: We don't exhaust the edge list as they claim
    # print(eord)
    # for k, v in eord.items():
    #     if v:
    #         pprint(eord)
    #         msg = f"{k} edge list was not exhausted"
    #         raise RuntimeError(msg)

    if len(seq) != len(new_seq):
        msg = f"Expected new_seq to be of len {len(seq)} but was {len(new_seq)}"
        raise RuntimeError(msg)
    if not same_klets(seq, new_seq, 2):
        msg = "Doublets are not preserved"
        raise RuntimeError(msg)
    if not same_klons(seq, new_seq, 3):
        msg = "Triplon are not preserved"
        raise RuntimeError(msg)
    # if not is_klet_preserved(seq, new_seq, 3):
    #     print("Ok, DtP did not preserve triplets")

    return new_seq


if __name__ == "__main__":
    # from Bio import SeqIO

    # path = "tests/test_data/test_rand_10000_genomic.fna"
    # for record in SeqIO.parse(path, "fasta"):
    # S1 = record.seq

    # print(S1)
    perm = doublet_preserving_permutation(S1)
    # print(perm, "DP")
    perm = doublet_and_triplet_preserving_permutation(S1)
    # print(perm, "DTP")
    perm = doublet_and_triplon_preserving_permutation(S1)
    print(perm, "DtP")

    print("Ok")

File: src/altschul.rs
use crate::unionfind::UnionFind;
use crate::utils::{each_cons, same_klets};
use rand::seq::SliceRandom;
use std::collections::{HashMap, HashSet};

pub const S1: &str = "AGACATAAAGTTCCGTACTGCCGGGAT";

fn check_if_connected(z_graph: &[&str], vertices: &[&str]) -> bool {
    debug_assert_eq!(z_graph.len(), vertices.len() - 1);

    let mut uf = UnionFind::new(vertices.len());
    for edge in z_graph {
        let fr = &edge[..edge.len() - 1];
        let to = &edge[1..];
        if let (Some(fr_idx), Some(to_idx)) = (
            vertices.iter().position(|v| *v == fr),
            vertices.iter().position(|v| *v == to),
        ) {
            uf.union(fr_idx, to_idx);
        } else {
            panic!();
        }
    }

    uf.group_count() == 1
}

fn edge_ordering(seq: &str, k: usize) -> HashMap<&str, Vec<&str>> {
    let mut eord: HashMap<&str, Vec<&str>> = HashMap::new();

    for chunk in each_cons(seq, k) {
        let key = &chunk[..k - 1];
        eord.entry(key).or_default().push(chunk);
    }

    // Ensure the last vertex is added
    let last_vertex = &seq[seq.len() - (k - 1)..];
    eord.entry(last_vertex).or_default();

    eord
}

pub fn klet_preserving_permutation(seq: &str, k: usize, debug: bool) -> String {
    assert!(1 < k);
    assert!(k < seq.len());

    let mut eord = edge_ordering(seq, k);
    let s1 = &seq[..k - 1];
    let sf = &seq[seq.len() - (k - 1)..];

    if debug {
        println!("1. {:?}", seq);
        println!("Edge ordering: {:?}", eord);
        println!("s1: {:?}, sf: {:?}", s1, sf);
    }

    let mut last_edges: Vec<&str> = Vec::new();
    let vertices: Vec<&str> = eord.keys().cloned().collect();
    let mut seen = HashSet::new();
    let mut max_iterations = 1000;

    while max_iterations > 0 {
        last_edges.clear();
        for (vertex, edges) in &eord {
            if *vertex != sf {
                if let Some(edge) = edges.choose(&mut rand::thread_rng()) {
                    last_edges.push(edge);
                }
            }
        }

        let key = last_edges.clone();
        if !seen.insert(key) {
            continue;
        }
        if check_if_connected(&last_edges, &vertices) {
            break;
        }
        max_iterations -= 1;
    }

    if max_iterations == 0 {
        panic!("Exhausted iterations. Returning original sequence.");
    }

    if debug {
        println!("5. {:?}", last_edges);
    }

    for last_edge in &last_edges {
        let vertex = &last_edge[..k - 1];
        if let Some(edges) = eord.get_mut(vertex) {
            if let Some(pos) = edges.iter().position(|e| e == last_edge) {
                edges.remove(pos);
                edges.push(last_edge);
            }
        }
    }

    let mut eord_perm = HashMap::new();
    for (vertex, mut edges) in eord {
        if vertex == sf {
            edges.shuffle(&mut rand::thread_rng());
            eord_perm.insert(vertex, edges);
        } else {
            let mut to_permute = edges[..edges.len() - 1].to_vec();
            to_permute.shuffle(&mut rand::thread_rng());
            to_permute.push(edges[edges.len() - 1]);
            eord_perm.insert(vertex, to_permute);
        }
    }

    let mut new_seq = String::from(s1);
    let mut cur_vertex = s1.to_string();
    loop {
        match eord_perm.get_mut(&cur_vertex[..]) {
            Some(edges) if !edges.is_empty() => {
                let cur_edge = edges.remove(0);
                cur_vertex = cur_edge[1..].to_string();
                new_seq.push(cur_edge.chars().last().unwrap());
                if debug {
                    println!("{:?}, {:?}, {:?}", cur_edge, cur_vertex, new_seq);
                }
            }
            _ => {
                assert_eq!(cur_vertex, sf);
                break;
            }
        }
    }

    for (key, value) in &eord_perm {
        if !value.is_empty() {
            panic!("{} edge list was not exhausted", key);
        }
    }

    assert_eq!(seq.len(), new_seq.len(), "Sequence length mismatch");
    assert!(same_klets(seq, &new_seq, 2), "Doublets are not preserved");

    new_seq
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_check_if_connected() {
        let z_graph = ["AT", "CT", "GA"];
        let vertices = ["A", "C", "G", "T"];
        assert!(check_if_connected(&z_graph, &vertices));
    }

    #[test]
    fn test_edge_ordering() {
        let result = edge_ordering(S1, 3);
        assert_eq!(result.len(), 16);
    }

    #[test]
    fn test_klet_preservation() {
        let _seq = "AGTACTAT".repeat(2);
        let seq = _seq.as_str();
        let seq_2 = klet_preserving_permutation(seq, 2, false);
        assert!(same_klets(seq, &seq_2, 2));

        let seq_3 = klet_preserving_permutation(seq, 3, false);
        assert!(same_klets(seq, &seq_3, 2));
        assert!(same_klets(seq, &seq_3, 3));
    }
}
